src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +64,246 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
 /* exports for SQL callable functions */
 PG_FUNCTION_INFO_V1(master_stage_shard_row);
 PG_FUNCTION_INFO_V1(master_stage_shard_placement_row);
+PG_FUNCTION_INFO_V1(citus_table_size);
+PG_FUNCTION_INFO_V1(citus_total_relation_size);
+PG_FUNCTION_INFO_V1(citus_relation_size);
+
+/*
+ * SupportFunctionForCitusSize is a helper function for all citus size functions.
+ *
+ */
+uint64
+SupportFunctionForCitusSize(Oid distributedRelationId, CitusSizeFunctionTypes
+							sizeFunctionType)
+{
+	DistTableCacheEntry *distributedTableCacheEntry = NULL;
+
+	/* Build shard qualified name. */
+	Oid schemaId = get_rel_namespace(distributedRelationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+
+	int shardIndex = 0;
+	uint64 totalRelationSize = 0;
+
+	MultiConnection *connection = NULL;
+
+	/* Check whether the size of table can be obtained with citus size functions. */
+	ErrorIfNotSuitableToGetSize(distributedRelationId);
+
+	distributedTableCacheEntry = DistributedTableCacheEntry(distributedRelationId);
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		char *workerNodeName = workerNode->workerName;
+		uint32 workerNodePort = workerNode->workerPort;
+		bool shardSeenOnNode = false;
+		PGresult *queryResult = NULL;
+		int executeCommand = 0;
+		char *tableSizeString;
+
+		StringInfo tableSizeQuery = makeStringInfo();
+		appendStringInfoString(tableSizeQuery, "Select ");
+
+		for (shardIndex = 0; shardIndex <
+			 distributedTableCacheEntry->shardIntervalArrayLength; shardIndex++)
+		{
+			ShardPlacement *placement =
+				distributedTableCacheEntry->arrayOfPlacementArrays[shardIndex];
+			char *shardNodeName = placement->nodeName;
+			uint32 shardNodePort = placement->nodePort;
+
+			/*Reference table exist on each node*/
+			if ((strcmp(shardNodeName, workerNodeName) == 0 && shardNodePort ==
+				 workerNodePort) ||
+				distributedTableCacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
```

Comment:
Do we need to check this? If a table is reference table, I think first part of the if statement will be true anyways. In fact this can cause adding counting size of a reference table's placement multiple times. Did I miss something?